src/rules/lowercase-name.ts

```diff
@@ -0,0 +1,129 @@
+import { createRule } from './tsUtils';
+import {
+  TSESTree,
+  AST_NODE_TYPES,
+} from '@typescript-eslint/experimental-utils';
+
+const isIdentifier = (
+  callee: TSESTree.LeftHandSideExpression,
+): callee is TSESTree.Identifier => !!callee;
+
+const isItTestOrDescribeFunction = (
+  callee: TSESTree.LeftHandSideExpression,
+): callee is TSESTree.Identifier => {
```

Comment:
Doesn't matter that much in this case, but maybe it helps someone learn more about TS: This is technically an incorrect type guard, because the function returns true not for any `TSESTree.Identifier`, but only for specific ones (those with certain names). This matters as soon as you negate the guard (`!isItTestOrDescribeFunction`) - consider the following code:
```ts
const node: Identifier | Literal = someIdentifier;
if (!isItTestOrDescribeFunction) {
  node; // Inferred as Literal
  node.value; // Typechecks even though it's really an identifier that has no `value`
}
```
Hope that this explanation is reasonably easy to understand :)

In this case I'd recommend removing the type check and making the signature
```ts
isItTestOrDescribeFunction = (callee: Identifier): boolean => callee.name // === ...
```
so callers have to do `isIdentifier(callee) && isItTestOrDescribeFunction(callee)`.

P.S. If you *really* need to type something like this correctly, I did it [here](https://github.com/facebook/jest/blob/15e2cc68150e5a67cc9f4b18995954696c9bbf59/packages/jest-config/src/utils.ts#L233-L238), but it's a bit arcane :D

Reply:
Yeah, I probably butchered the implementation here 😅 mostly kept the helpers the same, but trying to make tsc shut up. Super happy to make them better if you wanna give it a whack :)

Reply:
I've redone it now to what I _think_ are valid type guards